### DISCUSSION TITLE
SPI SD Card to SSD1306 12C OLED Display in an ESP32?

### DISCUSSION THREAD

**Discussion by cybergunk** [espressif/arduino-esp32#7777] [Q&A]
Hi! Apologies if this question is a bit uninformed in more ways than I can imagine, I am new to creating projects involving microcontrollers.

Am I able to communicate a peripheral that uses SPI pins like a [Micro SD card breakout board](https://www.amazon.com/risingsaplings-Module-Breakout-Interface-Conversion/dp/B09FK8QGH9/ref=sr_1_1_sspa?keywords=sd+card+breakout&qid=1675144598&sr=8-1-spons&psc=1&spLa=ZW5jcnlwdGVkUXVhbGlmaWVyPUExSVBGU0tWWDVKUjFGJmVuY3J5cHRlZElkPUEwMzY1NDk2MVo4WjlZWjJCMVg0USZlbmNyeXB0ZWRBZElkPUEwODk4NzE4SFlUNkVDTDhZM001JndpZGdldE5hbWU9c3BfYXRmJmFjdGlvbj1jbGlja1JlZGlyZWN0JmRvTm90TG9nQ2xpY2s9dHJ1ZQ==) to an 12C output like this [OLED Display Module](https://www.amazon.com/DIYmall-Serial-128x64-Display-Arduino/dp/B00O2KDQBE?th=1)? If yes, how do I pin these to an ESP32? More importantly how do I code them (using Micropython) so that the MicroSD communicates with the 12C?

I'm creating a project similar to this: [Play Video on a Tiny OLED Display/Matrix - ESP32 Project 7](https://www.youtube.com/watch?v=Bh1DAStx5yk&t=35s) and I understand that his OLED has SPI pins. I found a [tutorial ](https://www.youtube.com/watch?v=qNLvoSQCx60&t=377s) inputting a MicroSD but to an 12S audio output and I do not know if it is the same

I already read through these references regarding SPI and I2C communication but I do not know how to incorporate them together:
[MicroPython: OLED Display with ESP32 and ESP8266](https://randomnerdtutorials.com/micropython-oled-display-esp32-esp8266/)
[ESP32 SPI Communication](https://randomnerdtutorials.com/esp32-spi-communication-arduino/)
[ESP32 I2C Communication](https://randomnerdtutorials.com/esp32-i2c-communication-arduino-ide/)

Thank you!

**Comment by VojtechBartoska** [comment]
@P-R-O-C-H-Y Can you please help with answering this?

**Comment by P-R-O-C-H-Y** [comment]
Hi @cybergunk, at first I am not sure if I understood correctly. You can use both peripherals (SPI and I2C) at the same time as it can be connected to different pins. You cannot use SPI + I2C on same pins. About the coding, you are in ESP32 Arduino repository, which uses C  (C++) to code, so about Micropython you need to ask on different repositories / forums.

**Comment by Ramotny** [comment]
Hello all
Since few days I'm trying to get ESP32 (Devkit V1) work together with SDcard and OLED display. Both are connected parallel do VSPI (MOSI=23, MISO=19, SCLK=18) except CS which is 5 for LED and 15 for SDcard.
I use Adafruit_SSD1306 OLED library and SD+SPI ESP libraries. If only one device is used (removin other code with #ifdef) the both work fine (separately) even if hardware is still connected of both SD and OLED.
As soon as both are active the OLED work until SD.begin(). From that moment OLED does not react/work anymore.
So the hardware is ok it must be software problem or conflict.
Different CS pins tested (0,5,15) - all work same manner - as soon as SD.begin() is used the LED stops working otherwise works ok.
I might show here the code but it's nothing extraordinary.
Where should I seek the problem and it's solution?
(be patient with me it my first post on GitHub)
Gregory

**Reply by VaibhavSugandhi** [reply]
Hi @Ramotny 

It is solved by calling SPI.end() after SD.begin() function. 

This is how I did it, 

`// This is a part of long code, use it carefully.
if (!SD.begin()) {
    Serial.println("Card Mount Failed");
    SPI.end();
  }
  SPI.end();

  Wire.begin();
  delay(10);
  display.begin(SSD1306_SWITCHCAPVCC, SCREEN_ADDRESS);
// Display on OLED as per need !
// Later, when you want to write anything in SD card, then call SD.begin(); again & SPI.end(); at the end of operation.`

I hope this helps you someway. 
Good luck !
